src/drivers/px4fmu/fmu.cpp

```diff
@@ -820,29 +780,63 @@ PX4FMU::set_i2c_bus_clock(unsigned bus, unsigned clock_hz)
 void
 PX4FMU::subscribe()
 {
-	/* subscribe/unsubscribe to required actuator control groups */
-	uint32_t sub_groups = _groups_required & ~_groups_subscribed;
-	uint32_t unsub_groups = _groups_subscribed & ~_groups_required;
-	_poll_fds_num = 0;
+	// first clear everything
+	ScheduleClear();
+
+	for (auto control_sub : _control_subs) {
+		control_sub.unregister_callback();
+	}
+
+	// if subscribed to control group 0 or 1 then move to the rate_ctrl WQ
+	const bool sub_group_0 = (_groups_required & (1 << 0));
+	const bool sub_group_1 = (_groups_required & (1 << 1));
+
+	if (sub_group_0 || sub_group_1) {
+		if (!WorkItem::Init(px4::wq_configurations::rate_ctrl)) {
+			PX4_ERR("changing WQ failed");
+		}
+	}
+
+	/*
+	* Adjust actuator topic update rate to keep up with
+	* the highest servo update rate configured.
+	*
+	* We always mix at max rate; some channels may update slower.
+	*/
+	int max_rate = (_pwm_default_rate > _pwm_alt_rate) ? _pwm_default_rate : _pwm_alt_rate;
+
+	// oneshot
+	if ((_pwm_default_rate == 0) || (_pwm_alt_rate == 0)) {
+		max_rate = 1000;
+	}
+
+	// max interval 1 - 100 ms (10 - 1000Hz)
+	const int update_interval_in_us = math::constrain(1000000 / max_rate, 1000, 100000);
+	_current_update_rate = max_rate;
 
+	// subscribe to all required actuator control groups with max interval set
 	for (unsigned i = 0; i < actuator_controls_s::NUM_ACTUATOR_CONTROL_GROUPS; i++) {
-		if (sub_groups & (1 << i)) {
+		if (_groups_required & (1 << i)) {
 			PX4_DEBUG("subscribe to actuator_controls_%d", i);
-			_control_subs[i] = orb_subscribe(_control_topics[i]);
-		}
 
-		if (unsub_groups & (1 << i)) {
-			PX4_DEBUG("unsubscribe from actuator_controls_%d", i);
-			orb_unsubscribe(_control_subs[i]);
-			_control_subs[i] = -1;
-		}
+			if (!_control_subs[i].register_callback()) {
```

Comment:
There's some threadding issues here: since you are changing work-queues potentially, this could already trigger a concurrent update while you are still in here on the previous work queue.

Reply:
Thanks, reviewing the threading is one of the reasons I've been sitting on this PR. At the moment I've moved the `_groups_subscribed` update to before the new callback is registered, but let me think about a proper solution that isn't obtrusive.

Reply:
What do you think about using the CDev lock() and unlock() (PX4FMU is a CDev)? See the last push.  https://github.com/PX4/Firmware/pull/12224/commits/7d98c52a6de4fa15d6487e005c31b4444993422d

Reply:
The problem goes deeper and adding a lock in `subscribe` is not enough. I think the cleanest solution w/o adding more locking is to just move the `register_callback` as the very last thing you do in `run()`.

Then the `ScheduleNow` call from ioctl is even more problematic. You need to make sure:
- When switching work queues, the work item is removed from the old queue and then you do the switch - this needs to be atomic from the perspective of a `ScheduleNow` call.
- The switch also needs to happen as the very last thing in `run()`

Reply:
I'll take a closer look tonight when I have time, but all the PWM IOCTL operations are within the cdev lock, and changing WorkQueues does clear any queued WorkItems before joining the next.

We can also have the run return immediately on subscription change.

Reply:
Yes I saw that and the lock helps with some of the issues. There is still more:
- The rest of `subscribe` is still subject to race conditions.
- The solution looks too fragile to me. There are various implicit expectations so that a refactoring will easily break it.

If work queue changes is something that other modules will use as well, it needs to be simpler and safer to use. You could do that by requesting a wq change which is then executed after the run call for example.

Reply:
Yes I agree it's fragile, but I don't think it's going to be a common pattern, and I'm not even sure it's worth the trouble of bothering here.

I've moved the subscribe to the end of Run().

Reply:
Updated subscribe() to return early if the WQ changes. The actual subscription updates will now happen in the new thread. https://github.com/PX4/Firmware/pull/12224/commits/c98c6d085bc4e614741ce8f3b28311f1a025a703

Reply:
Ok, I'm good with these changes if you add some comments making the assumptions here explicit:
- the lock calls, why they are needed and what they protect against.
- the method docs of `WorkItem::Init`: the caller is responsible for synchronization, and returning immediately from run after the call.

Reply:
Comments added and tested again.